In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.model_selection import train_test_split, GridSearchCV 
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score 
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.metrics import accuracy_score

In [2]:
df = pd.read_csv('winemag-data-130k-v2 2.csv')
df.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

In [3]:
len(df)

129971

In [4]:
df.variety.describe()

count         129970
unique           707
top       Pinot Noir
freq           13272
Name: variety, dtype: object

In [5]:
nullvals = df.variety.isnull().values
len(nullvals)

129971

## Find out how many null values exist and get rid of the features with the most null vals

In [6]:

df.isna().sum()

Unnamed: 0                   0
country                     63
description                  0
designation              37465
points                       0
price                     8996
province                    63
region_1                 21247
region_2                 79460
taster_name              26244
taster_twitter_handle    31213
title                        0
variety                      1
winery                       0
dtype: int64

In [7]:
# region_2 and twitter handle have the most null values so dropping the column
df_cleaned1 = df.drop(['Unnamed: 0','region_2','taster_twitter_handle'],axis=1)
df_cleaned1

country                                        description  \
0             Italy  Aromas include tropical fruit, broom, brimston...   
1          Portugal  This is ripe and fruity, a wine that is smooth...   
2                US  Tart and snappy, the flavors of lime flesh and...   
3                US  Pineapple rind, lemon pith and orange blossom ...   
4                US  Much like the regular bottling from 2012, this...   
5             Spain  Blackberry and raspberry aromas show a typical...   
6             Italy  Here's a bright, informal red that opens with ...   
7            France  This dry and restrained wine offers spice in p...   
8           Germany  Savory dried thyme notes accent sunnier flavor...   
9            France  This has great depth of flavor with its fresh ...   
10               US  Soft, supple plum envelopes an oaky structure ...   
11           France  This is a dry wine, very spicy, with a tight, ...   
12               US  Slightly reduced, this wine offers a chalky, t...   
13            Italy  This is dominated by oak and oak-driven aromas...   
14               US  Building on 150 years and six generations of w...   
15          Germany  Zesty orange peels and apple notes abound in t...   
16        Argentina  Baked plum, molasses, balsamic vinegar and che...   
17        Argentina  Raw black-cherry aromas are direct and simple ...   
18            Spain  Desiccated blackberry, leather, charred wood a...   
19               US  Red fruit aromas pervade on the nose, with cig...   
20               US  Ripe aromas of dark berries mingle with ample ...   
21               US  A sleek mix of tart berry, stem and herb, alon...   
22            Italy  Delicate aromas recall white flower and citrus...   
23               US  This wine from the Geneseo district offers aro...   
24            Italy  Aromas of prune, blackcurrant, toast and oak c...   
25               US  Oak and earth intermingle around robust aromas...   
26            Italy  Pretty aromas of yellow flower and stone fruit...   
27            Italy  Aromas recall ripe dark berry, toast and a whi...   
28            Italy  Aromas suggest mature berry, scorched earth, a...   
29               US  Clarksburg is becoming a haven for Chenin Blan...   
...             ...                                                ...   
129941           US  A Chardonnay with an unusual companion, 8% Sém...   
129942           US  This is classic in herbaceous aromas and flavo...   
129943        Italy  A blend of Nero d'Avola and Syrah, this convey...   
129944       Israel  Deep garnet in the glass, this has a nose of b...   
129945           US  Hailing from one of the more popular vineyards...   
129946      Germany  Plump, clingy peach and honey notes are cut wi...   
129947        Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
129948    Argentina  Raspberry and cassis aromas are fresh and upri...   
129949           US  There's no bones about the use of oak in this ...   
129950           US  This opens with herbaceous dollops of thyme an...   
129951       France  Hugely spicy this rich wine is described as sw...   
129952           US  This Zinfandel from the eastern section of Nap...   
129953  New Zealand  Roughly two-thirds Cabernet and one-third Merl...   
129954  New Zealand  One of the more characterful Pinot Gris for th...   
129955  New Zealand  Like Dog Point's 2011 Chardonnay, this wine is...   
129956  New Zealand  The blend is 44% Merlot, 33% Cabernet Sauvigno...   
129957        Spain  Lightly baked berry aromas vie for attention w...   
129958  New Zealand  This blend of Cabernet Sauvignon-Merlot and Ca...   
129959       France  The granite soil of the Brand Grand Cru vineya...   
129960     Portugal  Fresh and fruity, this is full of red cherry f...   
129961        Italy  Intense aromas of wild cherry, baking spice, t...   
129962        Italy  Blackberry, cassis, grilled herb and toasted a...   
129963       Isr

In [8]:
df_cleaned2 = df_cleaned1.dropna()

In [9]:
len(df_cleaned2) 

54170

In [10]:
len(df) - len(df_cleaned2) #number of lost data points - lost over half our data points

75801

In [11]:
df_cleaned2

country                                        description  \
3              US  Pineapple rind, lemon pith and orange blossom ...   
4              US  Much like the regular bottling from 2012, this...   
5           Spain  Blackberry and raspberry aromas show a typical...   
6           Italy  Here's a bright, informal red that opens with ...   
9          France  This has great depth of flavor with its fresh ...   
10             US  Soft, supple plum envelopes an oaky structure ...   
16      Argentina  Baked plum, molasses, balsamic vinegar and che...   
17      Argentina  Raw black-cherry aromas are direct and simple ...   
18          Spain  Desiccated blackberry, leather, charred wood a...   
20             US  Ripe aromas of dark berries mingle with ample ...   
22          Italy  Delicate aromas recall white flower and citrus...   
23             US  This wine from the Geneseo district offers aro...   
24          Italy  Aromas of prune, blackcurrant, toast and oak c...   
25             US  Oak and earth intermingle around robust aromas...   
26          Italy  Pretty aromas of yellow flower and stone fruit...   
28          Italy  Aromas suggest mature berry, scorched earth, a...   
35             US  As with many of the Erath 2010 vineyard design...   
42         France  This is a festive wine, with soft, ripe fruit ...   
49         France  Soft and fruity, this is a generous, ripe wine...   
53         France  Fruity and lightly herbaceous, this has fine t...   
60             US  Syrupy and dense, this wine is jammy in plum a...   
61          Italy  This densely hued wine has aromas of black plu...   
62             US  The aromas are brooding, with notes of barrel ...   
63         France  This fat, yeasty Champagne is comprised predom...   
64             US  There are intriguing touches to the nose of th...   
67             US  A blend of Merlot and Cabernet Franc, this win...   
69         France  This cooperative, based in Aÿ, has benefited f...   
71             US  Big oak defines this robustly dense and extrac...   
72          Italy  Aromas of black-skinned fruit, leather, underb...   
73             US  Juicy plum, raspberry and pencil lead lead the...   
...           ...                                                ...   
129902      Spain  Dry apple and lemon aromas are clean and one s...   
129903     France  A fine value for a northern Rhône of this qual...   
129904         US  A blend of Sauvignon Blanc, Sauvignon Gris and...   
129909         US  Hailing from an extreme coastal vineyard, this...   
129912         US  From the Ranch House block of the famous viney...   
129913         US  This is taut and dense, and requires time and ...   
129915         US  The third vintage of Mith is the best yet, com...   
129917  Argentina  This dark, meaty Malbec has some Agrelo fruit ...   
129919         US  This ripe, rich, almost decadently thick wine ...   
129921  Argentina  There is a select group of under-$20 Malbecs f...   
129925  Argentina  A lively, well-made blend of Tempranillo, Malb...   
129926         US  This pure Syrah from Reininger's estate vineya...   
129928      Spain  This represents Torres' highest effort in Prio...   
129930      Spain  Arguably on the heavy and rich side, but if yo...   
129932  Argentina  Andeluna's top wines tend to be ripe and plump...   
129936     France  A blend from several premier cru vineyards in ...   
129938  Argentina  Compared to the regular 2006 Malbec from Chaka...   
129943      Italy  A blend of Nero d'Avola and Syrah, this convey...   
129945         US  Hailing from one of the more popular vineyards...   
129947      Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
129948  Argentina  Raspberry and cassis aromas are fresh and upri...   
129949         US  There's no bones about the use of oak in this ...   
129950         US  This opens with herbaceous dollops of thyme an...   
129951     France  Hugely spicy this rich wine is 

In [12]:
list(df_cleaned2.columns.values)

['country',
 'description',
 'designation',
 'points',
 'price',
 'province',
 'region_1',
 'taster_name',
 'title',
 'variety',
 'winery']

In [13]:
#double check no null vals
df_cleaned2.isna().sum()

country        0
description    0
designation    0
points         0
price          0
province       0
region_1       0
taster_name    0
title          0
variety        0
winery         0
dtype: int64

In [14]:
rwo_df = pd.read_excel('Types of Wine - O, R, W.xlsx')
rwo_df.head()

variety   color
0  Alvar Roxo  Orange
1       Ar110  Orange
2        Ar99  Orange
3  Barbarossa  Orange
4   Barbaroux  Orange

In [15]:
df_cleaned2.describe(include= 'all')

country                                        description designation  \
count    54170                                              54170       54170   
unique       7                                              49537       23943   
top         US  Seductively tart in lemon pith, cranberry and ...     Reserve   
freq     25888                                                  3        1159   
mean       NaN                                                NaN         NaN   
std        NaN                                                NaN         NaN   
min        NaN                                                NaN         NaN   
25%        NaN                                                NaN         NaN   
50%        NaN                                                NaN         NaN   
75%        NaN                                                NaN         NaN   
max        NaN                                                NaN         NaN   

              points         price    province              region_1  \
count   54170.000000  54170.000000       54170                 54170   
unique           NaN           NaN          63                  1004   
top              NaN           NaN  California  Columbia Valley (WA)   
freq             NaN           NaN       13998                  2488   
mean       88.966550     39.209341         NaN                   NaN   
std         2.955188     36.154076         NaN                   NaN   
min        80.000000      4.000000         NaN                   NaN   
25%        87.000000     20.000000         NaN                   NaN   
50%        89.000000     30.000000         NaN                   NaN   
75%        91.000000     48.000000         NaN                   NaN   
max       100.000000   2013.000000         NaN                   NaN   

       taster_name                                        title     variety  \
count        54170                                        54170       54170   
unique          17                                        48994         434   
top     Roger Voss  Segura Viudas NV Extra Dry Sparkling (Cava)  Pinot Noir   
freq          9237                                            8        6736   
mean           NaN                                          NaN         NaN   
std            NaN                                          NaN         NaN   
min            NaN                                          NaN         NaN   
25%            NaN                                          NaN         NaN   
50%            NaN                                          NaN         NaN   
75%            NaN                                          NaN         NaN   
max            NaN                                          NaN         NaN   

                winery  
count            54170  
unique            8583  
top     Columbia Crest  
freq               157  
mean               NaN  
std                NaN  
min                NaN  
25%                NaN  
50%                NaN  
75%                NaN  
max                NaN

In [16]:
df_cleaned3 = df_cleaned2.merge(rwo_df, on='variety', how='left')
df_cleaned3

country                                        description  \
0             US  Pineapple rind, lemon pith and orange blossom ...   
1             US  Much like the regular bottling from 2012, this...   
2          Spain  Blackberry and raspberry aromas show a typical...   
3          Italy  Here's a bright, informal red that opens with ...   
4         France  This has great depth of flavor with its fresh ...   
5             US  Soft, supple plum envelopes an oaky structure ...   
6      Argentina  Baked plum, molasses, balsamic vinegar and che...   
7      Argentina  Raw black-cherry aromas are direct and simple ...   
8          Spain  Desiccated blackberry, leather, charred wood a...   
9             US  Ripe aromas of dark berries mingle with ample ...   
10         Italy  Delicate aromas recall white flower and citrus...   
11            US  This wine from the Geneseo district offers aro...   
12         Italy  Aromas of prune, blackcurrant, toast and oak c...   
13            US  Oak and earth intermingle around robust aromas...   
14         Italy  Pretty aromas of yellow flower and stone fruit...   
15         Italy  Aromas suggest mature berry, scorched earth, a...   
16            US  As with many of the Erath 2010 vineyard design...   
17        France  This is a festive wine, with soft, ripe fruit ...   
18        France  Soft and fruity, this is a generous, ripe wine...   
19        France  Fruity and lightly herbaceous, this has fine t...   
20            US  Syrupy and dense, this wine is jammy in plum a...   
21         Italy  This densely hued wine has aromas of black plu...   
22            US  The aromas are brooding, with notes of barrel ...   
23        France  This fat, yeasty Champagne is comprised predom...   
24            US  There are intriguing touches to the nose of th...   
25            US  A blend of Merlot and Cabernet Franc, this win...   
26        France  This cooperative, based in Aÿ, has benefited f...   
27            US  Big oak defines this robustly dense and extrac...   
28         Italy  Aromas of black-skinned fruit, leather, underb...   
29            US  Juicy plum, raspberry and pencil lead lead the...   
...          ...                                                ...   
54146      Spain  Dry apple and lemon aromas are clean and one s...   
54147     France  A fine value for a northern Rhône of this qual...   
54148         US  A blend of Sauvignon Blanc, Sauvignon Gris and...   
54149         US  Hailing from an extreme coastal vineyard, this...   
54150         US  From the Ranch House block of the famous viney...   
54151         US  This is taut and dense, and requires time and ...   
54152         US  The third vintage of Mith is the best yet, com...   
54153  Argentina  This dark, meaty Malbec has some Agrelo fruit ...   
54154         US  This ripe, rich, almost decadently thick wine ...   
54155  Argentina  There is a select group of under-$20 Malbecs f...   
54156  Argentina  A lively, well-made blend of Tempranillo, Malb...   
54157         US  This pure Syrah from Reininger's estate vineya...   
54158      Spain  This represents Torres' highest effort in Prio...   
54159      Spain  Arguably on the heavy and rich side, but if yo...   
54160  Argentina  Andeluna's top wines tend to be ripe and plump...   
54161     France  A blend from several premier cru vineyards in ...   
54162  Argentina  Compared to the regular 2006 Malbec from Chaka...   
54163      Italy  A blend of Nero d'Avola and Syrah, this convey...   
54164         US  Hailing from one of the more popular vineyards...   
54165      Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
54166  Argentina  Raspberry and cassis aromas are fresh and upri...   
54167         US  There's no bones about the use of oak in this ...   
54168         US  This opens with herbaceous dollops of thyme an...   
54169     France  Hugely spicy this rich wine is described as sw...   
54170      Spain  Lightly baked b

In [17]:
df_cleaned3.isna().sum()

country           0
description       0
designation       0
points            0
price             0
province          0
region_1          0
taster_name       0
title             0
variety           0
winery            0
color          7828
dtype: int64

In [18]:
df_cleaned4 = df_cleaned3.dropna()
df_cleaned4.color.describe()

count     46348
unique        3
top         Red
freq      30919
Name: color, dtype: object

In [19]:
df_cleaned4.groupby('color').count()


country  description  designation  points  price  province  region_1  \
color                                                                          
Orange      655          655          655     655    655       655       655   
Red       30919        30919        30919   30919  30919     30919     30919   
White     14774        14774        14774   14774  14774     14774     14774   

        taster_name  title  variety  winery  
color                                        
Orange          655    655      655     655  
Red           30919  30919    30919   30919  
White         14774  14774    14774   14774

# Potential target variables to Predict based on data above:
- color: there are 3 unique colors
- country: there are 7 unique countries
- points - below mean or above mean?
- price - below mean or above mean?
- taster name - based on the language they used?
- the variety of wine? - pinot noir etc (there are 434 total wine varieties)

In [20]:
df_cleaned4.describe(include= 'all')

country                                        description designation  \
count    46348                                              46348       46348   
unique       7                                              42427       20508   
top         US  Cigar box, café au lait, and dried tobacco aro...     Reserve   
freq     23577                                                  3        1113   
mean       NaN                                                NaN         NaN   
std        NaN                                                NaN         NaN   
min        NaN                                                NaN         NaN   
25%        NaN                                                NaN         NaN   
50%        NaN                                                NaN         NaN   
75%        NaN                                                NaN         NaN   
max        NaN                                                NaN         NaN   

              points         price    province              region_1  \
count   46348.000000  46348.000000       46348                 46348   
unique           NaN           NaN          62                   906   
top              NaN           NaN  California  Columbia Valley (WA)   
freq             NaN           NaN       12732                  2263   
mean       89.086131     41.197808         NaN                   NaN   
std         2.966685     37.211296         NaN                   NaN   
min        80.000000      4.000000         NaN                   NaN   
25%        87.000000     20.000000         NaN                   NaN   
50%        89.000000     32.000000         NaN                   NaN   
75%        91.000000     50.000000         NaN                   NaN   
max       100.000000   2013.000000         NaN                   NaN   

       taster_name                                        title     variety  \
count        46348                                        46348       46348   
unique          17                                        41926         264   
top     Roger Voss  Segura Viudas NV Extra Dry Sparkling (Cava)  Pinot Noir   
freq          7385                                            8        6736   
mean           NaN                                          NaN         NaN   
std            NaN                                          NaN         NaN   
min            NaN                                          NaN         NaN   
25%            NaN                                          NaN         NaN   
50%            NaN                                          NaN         NaN   
75%            NaN                                          NaN         NaN   
max            NaN                                          NaN         NaN   

                winery  color  
count            46348  46348  
unique            7734      3  
top     Columbia Crest    Red  
freq               150  30919  
mean               NaN    NaN  
std                NaN    NaN  
min                NaN    NaN  
25%                NaN    NaN  
50%                NaN    NaN  
75%                NaN    NaN  
max                NaN    NaN

In [22]:
# Remove some columns:
df_cleaned5 = df_cleaned4.drop(['description', 'designation', 'region_1', 'title', 'variety','winery'], axis=1, inplace=False)
df_cleaned5.head()

country  points  price           province         taster_name   color
0      US      87   13.0           Michigan  Alexander Peartree   White
1      US      87   65.0             Oregon        Paul Gregutt     Red
2   Spain      87   15.0     Northern Spain   Michael Schachner     Red
3   Italy      87   16.0  Sicily & Sardinia       Kerin O’Keefe     Red
4  France      87   27.0             Alsace          Roger Voss  Orange

In [30]:
one_hot_df = pd.get_dummies(df_cleaned5)
one_hot_df.head()

points  price  country_Argentina  country_Australia  country_Canada  \
0      87   13.0                  0                  0               0   
1      87   65.0                  0                  0               0   
2      87   15.0                  0                  0               0   
3      87   16.0                  0                  0               0   
4      87   27.0                  0                  0               0   

   country_France  country_Italy  country_Spain  country_US  province_Alsace  \
0               0              0              0           1                0   
1               0              0              0           1                0   
2               0              0              1           0                0   
3               0              1              0           0                0   
4               1              0              0           0                1   

      ...       taster_name_Matt Kettmann  taster_name_Michael Schachner  \
0     ...                               0                              0   
1     ...                               0                              0   
2     ...                               0                              1   
3     ...                               0                              0   
4     ...                               0                              0   

   taster_name_Paul Gregutt  taster_name_Roger Voss  \
0                         0                       0   
1                         1                       0   
2                         0                       0   
3                         0                       0   
4                         0                       1   

   taster_name_Sean P. Sullivan  taster_name_Susan Kostrzewa  \
0                             0                            0   
1                             0                            0   
2                             0                            0   
3                             0                            0   
4                             0                            0   

   taster_name_Virginie Boone  color_Orange  color_Red  color_White  
0                           0             0          0            1  
1                           0             0          1            0  
2                           0             0          1            0  
3                           0             0          1            0  
4                           0             1          0            0  

[5 rows x 91 columns]

In [31]:
labels = one_hot_df.color_Red
one_hot_df.drop('color_Red', axis=1, inplace=True)
one_hot_df.head()

points  price  country_Argentina  country_Australia  country_Canada  \
0      87   13.0                  0                  0               0   
1      87   65.0                  0                  0               0   
2      87   15.0                  0                  0               0   
3      87   16.0                  0                  0               0   
4      87   27.0                  0                  0               0   

   country_France  country_Italy  country_Spain  country_US  province_Alsace  \
0               0              0              0           1                0   
1               0              0              0           1                0   
2               0              0              1           0                0   
3               0              1              0           0                0   
4               1              0              0           0                1   

      ...       taster_name_Lauren Buzzeo  taster_name_Matt Kettmann  \
0     ...                               0                          0   
1     ...                               0                          0   
2     ...                               0                          0   
3     ...                               0                          0   
4     ...                               0                          0   

   taster_name_Michael Schachner  taster_name_Paul Gregutt  \
0                              0                         0   
1                              0                         1   
2                              1                         0   
3                              0                         0   
4                              0                         0   

   taster_name_Roger Voss  taster_name_Sean P. Sullivan  \
0                       0                             0   
1                       0                             0   
2                       0                             0   
3                       0                             0   
4                       1                             0   

   taster_name_Susan Kostrzewa  taster_name_Virginie Boone  color_Orange  \
0                            0                           0             0   
1                            0                           0             0   
2                            0                           0             0   
3                            0                           0             0   
4                            0                           0             1   

   color_White  
0            1  
1            0  
2            0  
3            0  
4            0  

[5 rows x 90 columns]

In [32]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
#* Use the scaler's `.fit_transform()` method to create a scaled version of our dataset. 
scaled_data = scaler.fit_transform(one_hot_df)
scaled_df = pd.DataFrame(scaled_data, columns=one_hot_df.columns)
scaled_df.head()

/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/anaconda3/lib/python3.6/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


points     price  country_Argentina  country_Australia  country_Canada  \
0 -0.703193 -0.757784          -0.237666          -0.171083       -0.058486   
1 -0.703193  0.639657          -0.237666          -0.171083       -0.058486   
2 -0.703193 -0.704036          -0.237666          -0.171083       -0.058486   
3 -0.703193 -0.677162          -0.237666          -0.171083       -0.058486   
4 -0.703193 -0.381550          -0.237666          -0.171083       -0.058486   

   country_France  country_Italy  country_Spain  country_US  province_Alsace  \
0       -0.480580      -0.385883      -0.312105    0.982758        -0.173011   
1       -0.480580      -0.385883      -0.312105    0.982758        -0.173011   
2       -0.480580      -0.385883       3.204048   -1.017544        -0.173011   
3       -0.480580       2.591460      -0.312105   -1.017544        -0.173011   
4        2.080819      -0.385883      -0.312105   -1.017544         5.779993   

      ...       taster_name_Lauren Buzzeo  taster_name_Matt Kettmann  \
0     ...                       -0.059772                   -0.30396   
1     ...                       -0.059772                   -0.30396   
2     ...                       -0.059772                   -0.30396   
3     ...                       -0.059772                   -0.30396   
4     ...                       -0.059772                   -0.30396   

   taster_name_Michael Schachner  taster_name_Paul Gregutt  \
0                      -0.410758                 -0.385551   
1                      -0.410758                  2.593693   
2                       2.434526                 -0.385551   
3                      -0.410758                 -0.385551   
4                      -0.410758                 -0.385551   

   taster_name_Roger Voss  taster_name_Sean P. Sullivan  \
0               -0.435361                     -0.264943   
1               -0.435361                     -0.264943   
2               -0.435361                     -0.264943   
3               -0.435361                     -0.264943   
4                2.296947                     -0.264943   

   taster_name_Susan Kostrzewa  taster_name_Virginie Boone  color_Orange  \
0                    -0.056021                   -0.398662     -0.119728   
1                    -0.056021                   -0.398662     -0.119728   
2                    -0.056021                   -0.398662     -0.119728   
3                    -0.056021                   -0.398662     -0.119728   
4                    -0.056021                   -0.398662      8.352263   

   color_White  
0     1.461894  
1    -0.684044  
2    -0.684044  
3    -0.684044  
4    -0.684044  

[5 rows x 90 columns]

In [33]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(one_hot_df, labels, test_size=0.25)

In [34]:
from sklearn.neighbors import KNeighborsClassifier
clf1 = KNeighborsClassifier()
clf1.fit(X_train, y_train)
test_preds = clf1.predict(X_test)

In [35]:
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score

In [36]:
def print_metrics(labels, preds):
    print("Precision Score: {}".format(precision_score(labels, preds)))
    print("Recall Score: {}".format(recall_score(labels, preds)))
    print("Accuracy Score: {}".format(accuracy_score(labels, preds)))
    print("F1 Score: {}".format(f1_score(labels, preds)))
    
print_metrics(y_test, test_preds)

Precision Score: 0.9325191216382926
Recall Score: 0.9694754392715147
Accuracy Score: 0.9322516613446103
F1 Score: 0.9506382443564109


> The precision score tells us how often our model was correct when predicting that the wine is red. 

> The recall score tells us how many of the actual red wines our model made correct classifications for. 

> The accuracy score tells us the overall percentage of correct predictions made by the model, 

> and f1-score is the harmonic mean of precision and recall, which represents a "balanced" metric between the two. Overall, f1-score is the most informative about the performance of the model, followed by accuracy. For multicategorical models, accuracy is best.



In [37]:
# take in six parameters: X_train, y_train, X_test, and  y_test,min_k and max_k. 
#Set these to 1 and 25, by default
def find_best_k(X_train, y_train, X_test, y_test, min_k=1, max_k=25):

# Create two variables, best_k and best_score
    best_k = 0
    best_score = 0.0
    
# For each iteration:Iterate through every odd number between min_k and max_k + 1.
    for k in range(min_k, max_k+1, 2):
        
# Create a new KNN classifier, and set the n_neighbors parameter to the current value for k, as determined by our loop
        knn = KNeighborsClassifier(n_neighbors=k)

    #     Fit this classifier to the training data.
        knn.fit(X_train, y_train)
    
# Generate predictions for X_test using the fitted classifier.
        preds = knn.predict(X_test)
    
# Calculate the F1-score for these predictions.  
        f1 = f1_score(y_test, preds)
    
 # Compare this F1-score to best_score. If better, update best_score and best_k. 
        if f1 > best_score:
            best_k = k
            best_score = f1

            # Once it has checked every value for k, print out the best value for k and the F1-score it achieved.        
    print("Best Value for k: {}".format(best_k))
    print("F1-Score: {}".format(best_score))


In [38]:
find_best_k(X_train, y_train, X_test, y_test)

Best Value for k: 1
F1-Score: 0.9646012074992056
